File: binance4j-wallet/src/main/java/com/binance4j/wallet/client/WalletClient.java
package com.binance4j.wallet.client;

import java.util.Map;

import com.binance4j.core.client.RestClient;
import com.binance4j.core.param.FramedPaging;
import com.binance4j.core.param.Params;
import com.binance4j.core.param.TimeFrame;
import com.binance4j.wallet.dto.AccountSnapshotType;
import com.binance4j.wallet.param.AccountSnapshotParams;
import com.binance4j.wallet.param.AccountStatusParams;
import com.binance4j.wallet.param.ApiPermissionsParams;
import com.binance4j.wallet.param.ApiTradingStatusParams;
import com.binance4j.wallet.param.AssetDetailParams;
import com.binance4j.wallet.param.AssetDividendRecordParams;
import com.binance4j.wallet.param.CoinInformationParams;
import com.binance4j.wallet.param.ConvertibleAssetParams;
import com.binance4j.wallet.param.DepositAddressParams;
import com.binance4j.wallet.param.DepositHistoryParams;
import com.binance4j.wallet.param.DustLogParams;
import com.binance4j.wallet.param.DustTransferParams;
import com.binance4j.wallet.param.FastWithdrawSwitchParams;
import com.binance4j.wallet.param.FundingWalletParams;
import com.binance4j.wallet.param.TradeFeeParams;
import com.binance4j.wallet.param.WalletTransferHistoryParams;
import com.binance4j.wallet.param.WalletTransferParams;
import com.binance4j.wallet.param.WithdrawHistoryParams;
import com.binance4j.wallet.param.WithdrawParams;
import com.binance4j.wallet.request.DisableFastWithdrawSwitchRequest;
import com.binance4j.wallet.request.DustTransferRequest;
import com.binance4j.wallet.request.EnableFastWithdrawSwitchRequest;
import com.binance4j.wallet.request.GetAccountStatus;
import com.binance4j.wallet.request.GetAllCoinsInfoRequest;
import com.binance4j.wallet.request.GetApiPermissionsRequest;
import com.binance4j.wallet.request.GetApiTradingStatusRequest;
import com.binance4j.wallet.request.GetAssetDetailRequest;
import com.binance4j.wallet.request.GetAssetDividendRecordRequest;
import com.binance4j.wallet.request.GetConvertibleAssetsRequest;
import com.binance4j.wallet.request.GetDepositAddressRequest;
import com.binance4j.wallet.request.GetDepositHistoryRequest;
import com.binance4j.wallet.request.GetDustLogRequest;
import com.binance4j.wallet.request.GetFundingWalletRequest;
import com.binance4j.wallet.request.GetFuturesAccountSnapshotRequest;
import com.binance4j.wallet.request.GetMarginAccountSnapshotRequest;
import com.binance4j.wallet.request.GetSpotAccountSnapshotRequest;
import com.binance4j.wallet.request.GetSystemStatusRequest;
import com.binance4j.wallet.request.GetTradeFeeRequest;
import com.binance4j.wallet.request.GetTransferHistoryRequest;
import com.binance4j.wallet.request.GetWithdrawHistoryRequest;
import com.binance4j.wallet.request.WalletTransferRequest;
import com.binance4j.wallet.request.WithdrawRequest;

/**
 * The API client for the wallet endpoints
 * 
 * @see <a href=
 *      "https://binance-docs.github.io/apidocs/spot/en/#wallet-endpoints">Documentation</a>
 */
public class WalletClient extends RestClient<WalletMapping> {
	/**
	 * @param key    API public key.
	 * @param secret API secret key.
	 */
	public WalletClient(String key, String secret) {
		super(WalletMapping.class, key, secret);
	}

	/**
	 * Fetches the system status.
	 * 
	 * @return The request to execute.
	 */
	public GetSystemStatusRequest getSystemStatus() {
		return new GetSystemStatusRequest(service.getSystemStatus());
	}

	/**
	 * Gets the information of coins (available for deposit and withdraw) for user .
	 * 
	 * @return The request to execute.
	 */
	public GetAllCoinsInfoRequest getAllCoinsInfo() {
		return new GetAllCoinsInfoRequest(service.getAllCoinsInfo(new CoinInformationParams().toMap()));
	}

	/**
	 * Fetches the daily SPOT account snapshots.
	 * 
	 * @param timeinterval Search interval.
	 * @return The request to execute.
	 */
	public GetSpotAccountSnapshotRequest getSpotAccountSnapshot(TimeFrame timeinterval) {
		return new GetSpotAccountSnapshotRequest(service.getSpotAccountSnapshot(AccountSnapshotType.SPOT,
				Params.merge(new AccountSnapshotParams(), timeinterval)));
	}

	/**
	 * Fetches the last 5 daily SPOT account snapshots.
	 * 
	 * @return The request to execute.
	 */
	public GetSpotAccountSnapshotRequest getSpotAccountSnapshot() {
		return new GetSpotAccountSnapshotRequest(
				service.getSpotAccountSnapshot(AccountSnapshotType.SPOT, new AccountSnapshotParams().toMap()));
	}

	/**
	 * Fetches the daily MARGIN account snapshots.
	 * 
	 * @param timeinterval Search interval.
	 * @return The request to execute.
	 */
	public GetMarginAccountSnapshotRequest getMarginAccountSnapshot(TimeFrame timeinterval) {
		return new GetMarginAccountSnapshotRequest(service.getMarginAccountSnapshot(AccountSnapshotType.MARGIN,
				Params.merge(new AccountSnapshotParams(), timeinterval)));
	}

	/**
	 * Fetches the last 5 daily MARGIN account snapshots.
	 * 
	 * @return The request to execute.
	 */
	public GetMarginAccountSnapshotRequest getMarginAccountSnapshot() {
		return new GetMarginAccountSnapshotRequest(
				service.getMarginAccountSnapshot(AccountSnapshotType.MARGIN, new AccountSnapshotParams().toMap()));
	}

	/**
	 * Fetches the daily FUTURES account snapshots.
	 * 
	 * @param timeinterval Search interval.
	 * @return The request to execute.
	 */
	public GetFuturesAccountSnapshotRequest getFuturesAccountSnapshot(TimeFrame timeinterval) {
		return new GetFuturesAccountSnapshotRequest(service.getFuturesAccountSnapshot(AccountSnapshotType.FUTURES,
				Params.merge(new AccountSnapshotParams(), timeinterval)));
	}

	/**
	 * Fetches the last 5 daily FUTURES account snapshots.
	 * 
	 * @return The request to execute.
	 */
	public GetFuturesAccountSnapshotRequest getFuturesAccountSnapshot() {
		return new GetFuturesAccountSnapshotRequest(
				service.getFuturesAccountSnapshot(AccountSnapshotType.FUTURES, new AccountSnapshotParams().toMap()));
	}

	/**
	 * Disables fast withdraw switch under your account.
	 * 
	 * @return The request to execute.
	 */
	public DisableFastWithdrawSwitchRequest disableFastWithdrawSwitch() {
		return new DisableFastWithdrawSwitchRequest(
				service.disableFastWithdrawSwitch(new FastWithdrawSwitchParams().toMap()));
	}

	/**
	 * Enables fast withdraw switch under your account.
	 * 
	 * @return The request to execute.
	 */
	public EnableFastWithdrawSwitchRequest enableFastWithdrawSwitch() {
		return new EnableFastWithdrawSwitchRequest(
				service.enableFastWithdrawSwitch(new FastWithdrawSwitchParams().toMap()));
	}

	/**
	 * Submits a withdraw request.
	 * <p>
	 * If network not send, you can get {@code network} and {@code isDefault} in
	 * networkList of a coin in the response of
	 * {@link #getAllCoinsInfo()}
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public WithdrawRequest withdraw(WithdrawParams params) {
		return new WithdrawRequest(service.withdraw(params.toMap()));
	}

	/**
	 * Fetches the deposit history of one or multiple coins.
	 * <p>
	 * Please notice the default {@code startTime} and {@code endTime} to make sure
	 * that time interval is within 0-90 days.
	 * <p>
	 * If both {@code startTime</code> and <code>endTime} are sent,
	 * time between {@code startTime</code> and <code>endTime} must be less than 90
	 * days.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetDepositHistoryRequest getDepositHistory(DepositHistoryParams params) {
		return new GetDepositHistoryRequest(service.getDepositHistory(params.toMap()));
	}

	/**
	 * Fetches the last deposit history of all coins.
	 * 
	 * @return The request to execute.
	 */
	public GetDepositHistoryRequest getDepositHistory() {
		return getDepositHistory(new DepositHistoryParams());
	}

	/**
	 * Fetches the deposit history of one or multiple coins.
	 * <p>
	 * Please notice the default {@code startTime} and {@code endTime} to make sure
	 * that time interval is within 0-90 days.
	 * <p>
	 * If both {@code startTime</code> and <code>endTime} are sent,
	 * time between {@code startTime</code> and <code>endTime} must be less than 90
	 * days.
	 * 
	 * @param params Request params.
	 * @param paging Paging.
	 * @return The request to execute.
	 */
	public GetDepositHistoryRequest getDepositHistory(DepositHistoryParams params, FramedPaging paging) {
		return new GetDepositHistoryRequest(
				service.getDepositHistory(Params.merge(params.toMap(), paging.toMap(Map.of("page", "offset")))));
	}

	/**
	 * Fetches the last deposit history of all coins.
	 * 
	 * @param paging Paging.
	 * @return The request to execute.
	 */
	public GetDepositHistoryRequest getDepositHistory(FramedPaging paging) {
		return new GetDepositHistoryRequest(service.getDepositHistory(
				Params.merge(new DepositHistoryParams().toMap(), paging.toMap(Map.of("page", "offset")))));
	}

	/**
	 * Fetches the withdraw history of one or multiple coins.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetWithdrawHistoryRequest getWithdrawHistory(WithdrawHistoryParams params) {
		return new GetWithdrawHistoryRequest(service.getWithdrawHistory(params.toMap()));
	}

	/**
	 * Fetches the withdraw history of one or multiple coins.
	 * 
	 * @param params Request params.
	 * @param paging Paging.
	 * @return The request to execute.
	 */
	public GetWithdrawHistoryRequest getWithdrawHistory(WithdrawHistoryParams params, FramedPaging paging) {
		return new GetWithdrawHistoryRequest(
				service.getWithdrawHistory(Params.merge(params.toMap(), paging.toMap(Map.of("page", "offset")))));
	}

	/**
	 * Fetches the last withdraw history of all coins.
	 * 
	 * @return The request to execute.
	 */
	public GetWithdrawHistoryRequest getWithdrawHistory() {
		return getWithdrawHistory(new WithdrawHistoryParams((String) null));
	}

	/**
	 * Fetches deposit address
	 * <p>
	 * If {@code network} is not sent, return the coin.
	 * <p>
	 * You can get {@code network} and {@code isDefault} in {@code networkList} in
	 * the response of {@link #getAllCoinsInfo}
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetDepositAddressRequest getDepositAddress(DepositAddressParams params) {
		return new GetDepositAddressRequest(service.getDepositAddress(params.toMap()));
	}

	/**
	 * Fetches the account status detail.
	 * 
	 * @return The request to execute.
	 */
	public GetAccountStatus getAccountStatus() {
		return new GetAccountStatus(service.getAccountstatus(new AccountStatusParams().toMap()));
	}

	/**
	 * Fetches the account api trading status detail.
	 * 
	 * @return The request to execute.
	 */
	public GetApiTradingStatusRequest getApiTradingStatus() {
		return new GetApiTradingStatusRequest(service.getApiTradingStatus(new ApiTradingStatusParams().toMap()));
	}

	/**
	 * Fetches the dust transfer logs.
	 * 
	 * @return The request to execute.
	 */
	public GetDustLogRequest getDustLog() {
		return new GetDustLogRequest(service.getDustLog(new DustLogParams().toMap()));
	}

	/**
	 * Fetches the dust transfer logs.
	 * 
	 * @param timeFrame Search interval.
	 * @return The request to execute.
	 */
	public GetDustLogRequest getDustLog(TimeFrame timeFrame) {
		return new GetDustLogRequest(service.getDustLog(Params.merge(new DustLogParams(), timeFrame)));
	}

	/**
	 * Retrieves the assets convertible into BNB.
	 * 
	 * @return The request to execute.
	 */
	public GetConvertibleAssetsRequest getConvertibleAssets() {
		ConvertibleAssetParams params = new ConvertibleAssetParams();
		return new GetConvertibleAssetsRequest(service.getConvertibleAssets(params.toMap()));
	}

	/**
	 * Retrieves the assets convertible into BNB.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetConvertibleAssetsRequest getConvertibleAssets(ConvertibleAssetParams params) {
		return new GetConvertibleAssetsRequest(service.getConvertibleAssets(params.toMap()));
	}

	/**
	 * Convert dust assets to BNB.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public DustTransferRequest dustTransfer(DustTransferParams params) {
		return new DustTransferRequest(service.dustTransfer(params.toMap()));
	}

	/**
	 * Queries the dividend record of one or multiple assets.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetAssetDividendRecordRequest getAssetDividendRecord(AssetDividendRecordParams params) {
		return new GetAssetDividendRecordRequest(service.getAssetDividendRecord(params.toMap()));
	}

	/**
	 * Queries the last 20 dividend asset records.
	 * 
	 * @return The request to execute.
	 */
	public GetAssetDividendRecordRequest getAssetDividendRecord() {
		return getAssetDividendRecord(new AssetDividendRecordParams(null));
	}

	/**
	 * Queries the dividend record of one or multiple assets.
	 * 
	 * @param params    Request params.
	 * @param timeFrame Search interval.
	 * @return The request to execute.
	 */
	public GetAssetDividendRecordRequest getAssetDividendRecord(AssetDividendRecordParams params, TimeFrame timeFrame) {
		return new GetAssetDividendRecordRequest(service.getAssetDividendRecord(Params.merge(params, timeFrame)));
	}

	/**
	 * Queries the last 20 dividend asset records.
	 * 
	 * @param timeFrame Search interval.
	 * @return The request to execute.
	 */
	public GetAssetDividendRecordRequest getAssetDividendRecord(TimeFrame timeFrame) {
		return new GetAssetDividendRecordRequest(
				service.getAssetDividendRecord(Params.merge(new AssetDividendRecordParams(null), timeFrame)));
	}

	/**
	 * Fetches the details of an asset supported on Binance. Please get network and
	 * other deposit or withdraw details from
	 * {@link #getAllCoinsInfo()}.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetAssetDetailRequest getAssetDetail(AssetDetailParams params) {
		return new GetAssetDetailRequest(service.getAssetDetail(params.toMap()));
	}

	/**
	 * Fetches the details of all assets supported on Binance.
	 * 
	 * @return The request to execute.
	 */
	public GetAssetDetailRequest getAssetDetail() {
		return getAssetDetail(new AssetDetailParams(null));
	}

	/**
	 * Fetches the trade fee.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetTradeFeeRequest getTradeFee(TradeFeeParams params) {
		return new GetTradeFeeRequest(service.getTradeFee(params.toMap()));
	}

	/**
	 * Fetches the trade fee.
	 * 
	 * @return The request to execute.
	 */
	public GetTradeFeeRequest getTradeFee() {
		return getTradeFee(new TradeFeeParams(null));
	}

	/**
	 * Make a universal transfer.
	 * <p>
	 * You need to enable {@code Permits Universal Transfer} option for the api key
	 * which requests this endpoint.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public WalletTransferRequest transfer(WalletTransferParams params) {
		return new WalletTransferRequest(service.transfer(params.toMap()));
	}

	/**
	 * Fetches the user universal transfer history
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetTransferHistoryRequest getTransferHistory(WalletTransferHistoryParams params) {
		return new GetTransferHistoryRequest(service.getTransferHistory(params.toMap()));
	}

	/**
	 * Fetches the user universal transfer history
	 * 
	 * @param params Request params.
	 * @param paging Paging.
	 * @return The request to execute.
	 */
	public GetTransferHistoryRequest getTransferHistory(WalletTransferHistoryParams params, FramedPaging paging) {
		return new GetTransferHistoryRequest(service.getTransferHistory(
				Params.merge(params.toMap(), paging.toMap(Map.of("limit", "size", "page", "current")))));
	}

	/**
	 * Fetches the funding wallet asset balance
	 * <p>
	 * Currently supports querying the following business assets：Binance Pay,
	 * Binance Card, Binance Gift Card, Stock Token
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetFundingWalletRequest getFundingWallet(FundingWalletParams params) {
		return new GetFundingWalletRequest(service.getFundingWallet(params.toMap()));
	}

	/**
	 * Fetches the funding wallet asset balance.
	 * 
	 * @return The request to execute.
	 */
	public GetFundingWalletRequest getFundingWallet() {
		return getFundingWallet(new FundingWalletParams(null, null));
	}

	/**
	 * Fetches the API Key Permission.
	 * 
	 * @return The request to execute.
	 */
	public GetApiPermissionsRequest getApiPermissions() {
		return new GetApiPermissionsRequest(service.getApiPermissions(new ApiPermissionsParams().toMap()));
	}
}

File: wallet/src/main/kotlin/com/binance4j/wallet/client/WalletClient.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.binance4j.wallet.client

import com.binance4j.core.Request
import com.binance4j.core.client.RestClient
import com.binance4j.core.param.FramedPaging
import com.binance4j.core.param.Params.Companion.merge
import com.binance4j.core.param.TimeFrame
import com.binance4j.wallet.dto.*
import com.binance4j.wallet.param.*

/**
 * The API client for the wallet endpoints
 * @param key    API public key.
 * @param secret API secret key.
 * [Documentation](https://binance-docs.github.io/apidocs/spot/en/.wallet-endpoints)
 */
class WalletClient(key: String, secret: String) : RestClient<WalletMapping>(WalletMapping::class.java, key, secret) {
    /**
     * Fetches the system status.
     * @return The request to execute.
     */
    fun getSystemStatus(): Request<SystemStatus> = Request(service.getSystemStatus())

    /**
     * Gets the information of coins (available for deposit and withdraw) for user .
     * @return The request to execute.
     */
    fun getAllCoinsInfo(): Request<List<CoinInformation>> = Request(service.getAllCoinsInfo(CoinInformationParams().toMap()))

    /**
     * Fetches the daily SPOT account snapshots.
     * @param timeFrame Search interval.
     * @return The request to execute.
     */
    @JvmOverloads
    fun getSpotAccountSnapshot(timeFrame: TimeFrame = TimeFrame()): Request<SpotAccountSnapshotResponse> = Request(
        service.getSpotAccountSnapshot(AccountSnapshotType.SPOT, merge(AccountSnapshotParams(), timeFrame))
    )

    /**
     * Fetches the daily MARGIN account snapshots.
     * @param timeFrame Search interval.
     * @return The request to execute.
     */
    fun getMarginAccountSnapshot(timeFrame: TimeFrame = TimeFrame()): Request<MarginAccountSnapshotResponse> {
        return Request(
            service.getMarginAccountSnapshot(AccountSnapshotType.MARGIN, merge(AccountSnapshotParams(), timeFrame))
        )
    }

    /**
     * Fetches the daily FUTURES account snapshots.
     * @param timeFrame Search interval.
     * @return The request to execute.
     */
    fun getFuturesAccountSnapshot(timeFrame: TimeFrame = TimeFrame()): Request<FuturesAccountSnapshotResponse> {
        return Request(
            service.getFuturesAccountSnapshot(AccountSnapshotType.FUTURES, merge(AccountSnapshotParams(), timeFrame))
        )
    }

    /**
     * Disables fast withdraw switch under your account.
     * @return The request to execute.
     */
    fun disableFastWithdrawSwitch(): Request<Unit> = Request(service.disableFastWithdrawSwitch(FastWithdrawSwitchParams().toMap()))

    /**
     * Enables fast withdraw switch under your account.
     * @return The request to execute.
     */
    fun enableFastWithdrawSwitch(): Request<Unit> = Request(service.enableFastWithdrawSwitch(FastWithdrawSwitchParams().toMap()))

    /**
     * Submits a withdrawal request. If network not send, you can get `network` and `isDefault` in networkList of a coin in the response of [getAllCoinsInfo]
     * @param params Request params.
     * @return The request to execute.
     */
    fun withdraw(params: WithdrawParams): Request<WithdrawResult> = Request(service.withdraw(params.toMap()))

    /**
     * Fetches the deposit history of one or multiple coins.
     * @param params Request params.
     * @return The request to execute.
     */
    @JvmOverloads
    fun getDepositHistory(params: DepositHistoryParams = DepositHistoryParams(), paging: FramedPaging = FramedPaging()): Request<List<DepositHistory>> =
        Request(service.getDepositHistory(merge(params, paging)))

    /**
     * Fetches the withdrawal history of one or multiple coins.
     * @param params Request params.
     * @return The request to execute.
     */
    @JvmOverloads
    fun getWithdrawHistory(
        params: WithdrawHistoryParams = WithdrawHistoryParams("", "", ""),
        paging: FramedPaging = FramedPaging()
    ): Request<List<WithdrawHistory>> =
        Request(service.getWithdrawHistory(merge(params, paging)))

    /**
     * Fetches deposit address
     * @param params Request params.
     * @return The request to execute.
     */
    fun getDepositAddress(params: DepositAddressParams): Request<DepositAddress> = Request(service.getDepositAddress(params.toMap()))

    /**
     * Fetches the account status detail.
     * @return The request to execute.
     */
    fun getAccountStatus(): Request<AccountStatus> = Request(service.getAccountstatus(AccountStatusParams().toMap()))

    /**
     * Fetches the account api trading status detail.
     * @return The request to execute.
     */
    fun getApiTradingStatus(): Request<ApiTradingStatus> = Request(service.getApiTradingStatus(ApiTradingStatusParams().toMap()))

    /**
     * Fetches the dust transfer logs.
     * @param timeFrame Search interval.
     * @return The request to execute.
     */
    @JvmOverloads
    fun getDustLog(timeFrame: TimeFrame = TimeFrame()): Request<DustLog> = Request(service.getDustLog(merge(DustLogParams(), timeFrame)))

    /**
     * Retrieves the assets convertible into BNB.
     * @return The request to execute.
     */
    fun getConvertibleAssets(): Request<ConvertibleAssets> = Request(service.getConvertibleAssets(ConvertibleAssetParams().toMap()))

    /**
     * Retrieves the assets convertible into BNB.
     * @param params Request params.
     * @return The request to execute.
     */
    fun getConvertibleAssets(params: ConvertibleAssetParams): Request<ConvertibleAssets> = Request(service.getConvertibleAssets(params.toMap()))

    /**
     * Convert dust assets to BNB.
     * @param params Request params.
     * @return The request to execute.
     */
    fun dustTransfer(params: DustTransferParams): Request<DustTransferResponse> = Request(service.dustTransfer(params.toMap()))

    /**
     * Queries the dividend record of one or multiple assets.
     * @param params Request params.
     * @return The request to execute.
     */
    fun getAssetDividendRecord(
        params: AssetDividendRecordParams = AssetDividendRecordParams(), timeFrame: TimeFrame = TimeFrame()
    ): Request<AssetDividendRecord> = Request(service.getAssetDividendRecord(merge(params, timeFrame)))

    /**
     * Fetches the details of an asset supported on Binance. Please get network and
     * other deposit or withdraw details from [getAllCoinsInfo].
     * @param params Request params.
     * @return The request to execute.
     */
    fun getAssetDetail(params: AssetDetailParams = AssetDetailParams()): Request<Map<String, AssetDetail>> = Request(service.getAssetDetail(params.toMap()))

    /**
     * Fetches the trade fee.
     * @param params Request params.
     * @return The request to execute.
     */
    fun getTradeFee(params: TradeFeeParams = TradeFeeParams()): Request<List<TradeFee>> = Request(service.getTradeFee(params.toMap()))

    /**
     * Make a universal transfer.
     * @param params Request params.
     * @return The request to execute.
     */
    fun transfer(params: WalletTransferParams): Request<WalletTransferResponse> = Request(service.transfer(params.toMap()))

    /**
     * Fetches the user universal transfer history
     * @param params Request params.
     * @return The request to execute.
     */
    fun getTransferHistory(params: WalletTransferHistoryParams, paging: FramedPaging = FramedPaging()): Request<WalletTransferHistory> =
        Request(service.getTransferHistory(merge(params.toMap(), paging.toMap(java.util.Map.of("limit", "size", "page", "current")))))


    /**
     * Fetches the funding wallet asset balance
     * @param params Request params.
     * @return The request to execute.
     */
    fun getFundingWallet(params: FundingWalletParams = FundingWalletParams()): Request<List<FundingAsset>> = Request(service.getFundingWallet(params.toMap()))

    /**
     * Fetches the API Key Permission.
     * @return The request to execute.
     */
    fun getApiPermissions(): Request<ApiPermissions> = Request(service.getApiPermissions(ApiPermissionsParams().toMap()))
}